models/trainModel.ts

```diff
@@ -0,0 +1,73 @@
+import pool from "../config/db";
+
+interface Train {
+  train_name: string;
+  train_number: string;
+  arrival_time: string;
+  departure_time: string;
+  date: string;
+  train_class: string;
+  train_type: string;
+  days_of_running: string;
+  quota: string;
+  travel_time: string;
+  fare: number;
+}
+
+const getStationByName = async (
+  stationName: string
+): Promise<string | null> => {
+  try {
+    const result = await pool.query(
+      "SELECT * FROM stations WHERE LOWER(station_name) LIKE LOWER($1)",
+      [`%${stationName}%`]
+    );
+    return result.rows[0] ? result.rows[0].id : null;
+  } catch (error) {
+    console.error("Error fetching station by name:", error);
+    throw new Error("Error fetching station by name");
+  }
+};
+
+const getTrainsByCities = async (
+  sourceStationName: string,
+  destinationStationName: string
+): Promise<Train[]> => {
+  try {
+    const sourceStationId = await getStationByName(sourceStationName);
+    const destinationStationId = await getStationByName(destinationStationName);

```

Comment:
Instead of calling the same two functions, you can create a single function that fetches both the source and destination station IDs from their name at once. that will reduce query calls